lib/Publisher.js

```diff
@@ -213,110 +206,60 @@ class Publisher extends EventEmitter {
-          // handle any chunking
-          let deserializeStream = new DeserializeStream();
-          subscriber.pipe(deserializeStream);
-          deserializeStream.on('message', subscriber.$handshake);
-
-          // if this publisher had the tcpNoDelay option set
-          // disable the nagle algorithm
-          if  (this._tcpNoDelay) {
-            subscriber.setNoDelay(true);
-          }
+  handleSubscriberConnection(subscriber, deserializeStream, header) {
+    let error = TcprosUtils.validateSubHeader(
+      header, this.getTopic(), this.getType(),
+      this._messageHandler.md5sum());
+    if (error !== null) {
+      this._log.error('Unable to validate connection header '
+                      + JSON.stringify(header));
+      subscriber.end(Serialize(error));
+      return;
+    }
+    // else
+    this._log.debug('Pub %s got connection header %s', this.getTopic(), JSON.stringify(header));
+
+    // create and send response
+    let respHeader =
+      TcprosUtils.createPubHeader(
+        this._nodeHandle.getNodeName(),
+        this._messageHandler.md5sum(),
+        this.getType(),
+        this.getLatching());
+    subscriber.write(respHeader);
+
+    // remove connections to deserializeStream - it's unnecessary now
+    deserializeStream.removeAllListeners();
+
+    // if this publisher had the tcpNoDelay option set
+    // disable the nagle algorithm
+    if  (this._tcpNoDelay) {
+      subscriber.setNoDelay(true);
+    }
 
-          subscriber.on('close', () => {
-            this._log.info('Publisher ' + this.getTopic() + ' client ' 
-                           + subscriber.name + ' disconnected!');
-            delete this._subClients[subscriber.name];
-          });
-
-          subscriber.on('end', () => {
-            this._log.info('Sub %s sent END', subscriber.name);
-          });
-
-          subscriber.on('error', () => {
-            this._log.info('Sub %s had error', subscriber.name);
-          });
-        }).listen(port);
-
-        // it's possible the port was taken before we could use it
-        server.on('error', (err) => {
-          if (err.code === 'EADDRINUSE') {
-            _createServer(callback);
-          }
-        });
+    subscriber.on('close', () => {
+      this._log.info('Publisher %s' + this.getTopic() + ' client '
+                     + subscriber.name + ' disconnected!');
+      delete this._subClients[subscriber.name];
+    });
 
-        // the port was available
-        server.on('listening', () => {
-          this._log.debug('Listening on port ' + port);
-          this._port = port;
-          this._server = server;
-          callback(port);
-        });
-      });
-    };
+    subscriber.on('end', () => {
+      this._log.info('Sub %s sent END', subscriber.name);
+    });
 
-    return new Promise((resolve, reject) => {
-      _createServer(resolve);
+    subscriber.on('error', () => {
+      this._log.warn('Sub %s had error', subscriber.name);
```

Comment:
Seems a bit funny to me that this error is reported as a warning. Was that intended?